File: WrapperCielo24/WrapperCielo24/Options.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Text.RegularExpressions;
using WrapperCielo24.JSON;

namespace WrapperCielo24
{
    /* The base class. All of the other option classes inherit from it. */
    public abstract class Options
    {
        public virtual Dictionary<string, string> GetDictionary()
        {
            Dictionary<string, string> queryDictionary = new Dictionary<string, string>();
            Type type = this.GetType();
            PropertyInfo[] properties = type.GetProperties();
            foreach (PropertyInfo property in properties)
            {
                QueryName key = (QueryName)property.GetCustomAttributes(typeof(QueryName), true).First();
                object value = property.GetValue(this, null);
                queryDictionary.Add(key.Name, GetStringValue(value));
            }
            return queryDictionary;
        }

        public virtual string ToQuery()
        {
            Dictionary<string, string> queryDictionary = this.GetDictionary();
            return Utils.ToQuery(queryDictionary);
        }

        /* Converts 'value' into string based on its type */
        protected string GetStringValue(object value)
        {
            if (value is List<string>)      // List<string> (returned as ["item", "item"]
            {
                List<string> stringList = ((List<string>)value);
                for (int i = 0; i < stringList.Count; i++)
                {
                    stringList[i] = "\"" + stringList[i] + "\""; // Add quotation marks
                }
                return "[" + String.Join(", ", stringList) + "]";
            }
            else if (value is List<Tag>)    // List<Tag> (returned as ["TAG", "TAG", "TAG"]
            {
                List<string> stringList = new List<string>();
                List<Tag> tagList = ((List<Tag>)value);
                for (int i = 0; i < tagList.Count; i++)
                {
                    stringList.Add("\"" + tagList[i].ToString() + "\""); // Add quotation marks
                }
                return "[" + String.Join(", ", stringList) + "]";
            }
            else if (value is List<Fidelity>)    // List<Fidelity> (returned as ["TAG", "TAG", "TAG"]
            {
                List<string> stringList = new List<string>();
                List<Fidelity> fidelityList = ((List<Fidelity>)value);
                for (int i = 0; i < fidelityList.Count; i++)
                {
                    stringList.Add("\"" + fidelityList[i].ToString() + "\""); // Add quotation marks
                }
                return "[" + String.Join(", ", stringList) + "]";
            }
            else if (value is char[])       // char[] (returned as (a, b))
            {
                return "(" + String.Join(", ", ((char[])value)) + ")";
            }
            else if (value is DateTime)     // DateTime (in ISO 8601 format)
            {
                DateTime dt = (DateTime)value;
                return (dt == null) ? "" : dt.ToString("yyyy-MM-ddTHH:mm:ss.fffffffzzz");
            }
            else                            // Takes care of the rest: int, bool, string, Uri
            {
                return (value == null) ? "" : value.ToString();
            }
        }
    }

    /* Options found in both Transcript and Caption options */
    /* All properties are nullable and are null by default */
    /* If the property is null during the ToQuery() method call, it is left out from the resulting query string */
    public abstract class CommonOptions : Options
    {
        [QueryName("characters_per_caption_line")]
        public int? CharactersPerCaptionLine { get; set; }
        [QueryName("elementlist_version")]
        public DateTime? ElementListVersion { get; set; }
        [QueryName("emit_speaker_change_token_as")]
        public string SpeakerChangeToken { get; set; }
        [QueryName("mask_profanity")]
        public bool? MaskProfanity { get; set; }
        [QueryName("remove_sounds_list")]
        public List<Tag> RemoveSoundsList { get; set; }
        [QueryName("remove_sound_references")]
        public bool? RemoveSoundReferences { get; set; }
        [QueryName("replace_slang")]
        public bool? ReplaceSlang { get; set; }
        [QueryName("sound_boundaries")]
        public char[] SoundBoundaries { get; set; }

        public CommonOptions(int? charactersPerCaptionLine = null,
                            DateTime? elementListVersion = null,
                            string speakerChangeToken = null,
                            bool? maskProfanity = null,
                            List<Tag> removeSoundsList = null,
                            bool? removeSoundReferences = null,
                            bool? replaceSlang = null,
                            char[] soundBoundaries = null)
        {
            this.CharactersPerCaptionLine = charactersPerCaptionLine;
            this.ElementListVersion = elementListVersion;
            this.SpeakerChangeToken = speakerChangeToken;
            this.MaskProfanity = maskProfanity;
            this.RemoveSoundsList = removeSoundsList;
            this.RemoveSoundReferences = removeSoundReferences;
            this.ReplaceSlang = replaceSlang;
            this.SoundBoundaries = soundBoundaries;
        }

        public virtual void FromQuery(string queryString)
        {
            Dictionary<string, string> dictionary = Regex.Matches(queryString, "([^?=&]+)(=([^&]*))?").Cast<Match>().ToDictionary(x => x.Groups[1].Value, x => x.Groups[3].Value);
            Type type = this.GetType();
            PropertyInfo[] properties = type.GetProperties();
            foreach (PropertyInfo property in properties)
            {
                QueryName key = (QueryName)property.GetCustomAttributes(false).First();
                if (dictionary.ContainsKey(key.Name))
                {
                    property.SetValue(this, dictionary[key.Name], null); // TODO convert to proper type
                }
            }
        }
    }

    /* All properties are nullable and are null by default */
    /* If the property is null during the ToQuery() method call, it is left out from the resulting query string */
    public class TranscriptOptions : CommonOptions
    {
        [QueryName("create_paragraphs")]
        public bool? CreateParagraphs { get; set; }
        [QueryName("newlines_after_paragraph")]
        public int? NewLinesAfterParagraph { get; set; }
        [QueryName("newlines_after_sentence")]
        public int? NewLinesAfterSentence { get; set; }
        [QueryName("timecode_every_paragraph")]
        public bool? TimeCodeEveryParagraph { get; set; }
        [QueryName("timecode_format")]
        public string TimeCodeFormat { get; set; }
        [QueryName("time_code_interval")]
        public int? TimeCodeInterval { get; set; }
        [QueryName("timecode_offset")]
        public int? TimeCodeOffset { get; set; }

        public TranscriptOptions(int? charactersPerCaptionLine = null,
                                DateTime? elementListVersion = null,
                                string speakerChangeToken = null,
                                bool? maskProfanity = null,
                                List<Tag> removeSoundsList = null,
                                bool? removeSoundReferences = null,
                                bool? replaceSlang = null,
                                char[] soundBoundaries = null,
                                bool? createParagraphs = null,
                                int? newLinesAfterParagraph = null,
                                int? newLinesAfterSentence = null,
                                bool? timecodeEveryParagraph = null,
                                string timecodeFormat = null,
                                int? timecodeInterval = null,
                                int? timecodeOffset = null)
            : base(charactersPerCaptionLine, elementListVersion, speakerChangeToken, maskProfanity, removeSoundsList, removeSoundReferences, replaceSlang, soundBoundaries)
        {
            this.CreateParagraphs = createParagraphs;
            this.NewLinesAfterParagraph = newLinesAfterParagraph;
            this.NewLinesAfterSentence = newLinesAfterSentence;
            this.TimeCodeEveryParagraph = timecodeEveryParagraph;
            this.TimeCodeFormat = timecodeFormat;
            this.TimeCodeInterval = timecodeInterval;
            this.TimeCodeOffset = timecodeOffset;
        }
    }

    /* All properties are nullable and are null by default */
    /* If the property is null during the ToQuery() method call, it is left out from the resulting query string */
    public class CaptionOptions : CommonOptions
    {
        [QueryName("build_url")]
        public bool? BuildUrl { get; set; }
        [QueryName("caption_words_min")]
        public int? CaptionWordsMin { get; set; }
        [QueryName("caption_by_sentence")]
        public bool? CaptionBySentence { get; set; }
        [QueryName("dfxp_header")]
        public string DfxpHeader { get; set; }
        [QueryName("disallow_dangling")]
        public bool? DisallowDangling { get; set; }
        [QueryName("display_effects_speaker_as")]
        public string EffectsSpeaker { get; set; }
        [QueryName("display_speaker_id")]
        public SpeakerId? DisplayedSpeakerId { get; set; }
        [QueryName("force_case")]
        public Case? ForceCase { get; set; }
        [QueryName("include_dfxp_metadata")]
        public bool? IncludeDfxpMetadata { get; set; }
        [QueryName("layout_target_caption_length_ms")]
        public int? LayoutTargetCaptionLengthMs { get; set; }
        [QueryName("line_break_on_sentence")]
        public bool? LineBreakOnSentence { get; set; }
        [QueryName("line_ending_format")]
        public LineEnding? LineEndingFormat { get; set; }
        [QueryName("lines_per_caption")]
        public int? LinesPerCaption { get; set; }
        [QueryName("maximum_caption_duration")]
        public int? MaximumCaptionDuration { get; set; }
        [QueryName("merge_gap_interval")]
        public int? MergeGapInterval { get; set; }
        [QueryName("minimum_caption_length_ms")]
        public int? MinimumCaptionLengthMs { get; set; }
        [QueryName("minimum_gap_between_captions_ms")]
        public int? MinimumGapBetweenCaptionsMs { get; set; }
        [QueryName("minimum_merge_gap_interval")]
        public int? MinimumMergeGapInterval { get; set; }
        [QueryName("qt_seamless")]
        public bool? QtSeamless { get; set; }
        [QueryName("silence_max_ms")]
        public int? SilenceMaxMs { get; set; }
        [QueryName("single_speaker_per_caption")]
        public bool? SingleSpeakerPerCaption { get; set; }
        [QueryName("sound_threshold")]
        public int? SoundThreshold { get; set; }
        [QueryName("sound_tokens_by_caption")]
        public bool? SoundTokensByCaption { get; set; }
        [QueryName("sound_tokens_by_line")]
        public bool? SoundTokensByLine { get; set; }
        [QueryName("sound_tokens_by_caption_list")]
        public List<Tag> SoundTokensByCaptionList { get; set; }
        [QueryName("sound_tokens_by_line_list")]
        public List<Tag> SoundTokensByLineList { get; set; }
        [QueryName("speaker_on_new_line")]
        public bool? SpeakerOnNewLine { get; set; }
        [QueryName("srt_format")]
        public string SrtFormat { get; set; }
        [QueryName("srt_print")]
        public bool? SrtPrintCaptionNumbers { get; set; }
        [QueryName("strip_square_brackets")]
        public bool? StripSquareBrackets { get; set; }
        [QueryName("utf8_mark")]
        public bool? Utf8Mark { get; set; }

        public CaptionOptions(int? charactersPerCaptionLine = null,
                                DateTime? elementListVersion = null,
                                string speakerChangeToken = null,
                                bool? maskProfanity = null,
                                List<Tag> removeSoundsList = null,
                                bool? removeSoundReferences = null,
                                bool? replaceSlang = null,
                                char[] soundBoundaries = null,
                                bool? buildUri = null,
                                int? captionWordsMin = null,
                                bool? captionBySentence = null,
                                string dfxpHeader = null,
                                bool? disallowDangling = null,
                                string effectsSpeaker = null,
                                SpeakerId? displaySpeakerId = null,
                                Case? forceCase = null,
                                bool? includeDfxpMetadata = null,
                                int? layoutDefaultCaptionLengthMs = null,
                                bool? lineBreakOnSentence = null,
                                LineEnding? lineEndingFormat = null,
                                int? linesPerCaption = null,
                                int? maximumCaptionDuration = null,
                                int? mergeGapInterval = null,
                                int? minimumCaptionLengthMs = null,
                                int? minimumGapBetweenCaptionsMs = null,
                                int? minimumMergeGapInterval = null,
                                bool? qtSeamless = null,
                                int? silenceMaxMs = null,
                                bool? singleSpeakerPerCaption = null,
                                int? soundThreshold = null,
                                bool? soundTokensByCaption = null,
                                bool? soundTokensByLine = null,
                                List<Tag> soundTokensByCaptionList = null,
                                List<Tag> soundTokensByLineList = null,
                                bool? speakerOnNewLine = null,
                                string srtFormat = null,
                                bool? srtPrintCaptionNumbers = null,
                                bool? stripSquareBrackets = null,
                                bool? utf8_mark = null)
            : base(charactersPerCaptionLine, elementListVersion, speakerChangeToken, maskProfanity, removeSoundsList, removeSoundReferences, replaceSlang, soundBoundaries)
        {
            this.BuildUrl = buildUri;
            this.CaptionWordsMin = captionWordsMin;
            this.CaptionBySentence = captionBySentence;
            this.DfxpHeader = dfxpHeader;
            this.DisallowDangling = disallowDangling;
            this.EffectsSpeaker = effectsSpeaker;
            this.DisplayedSpeakerId = displaySpeakerId;
            this.ForceCase = forceCase;
            this.IncludeDfxpMetadata = includeDfxpMetadata;
            this.LayoutTargetCaptionLengthMs = layoutDefaultCaptionLengthMs;
            this.LineBreakOnSentence = lineBreakOnSentence;
            this.LineEndingFormat = lineEndingFormat;
            this.LinesPerCaption = linesPerCaption;
            this.MaximumCaptionDuration = maximumCaptionDuration;
            this.MergeGapInterval = mergeGapInterval;
            this.MinimumCaptionLengthMs = minimumCaptionLengthMs;
            this.MinimumGapBetweenCaptionsMs = minimumGapBetweenCaptionsMs;
            this.MinimumMergeGapInterval = minimumMergeGapInterval;
            this.QtSeamless = qtSeamless;
            this.SilenceMaxMs = silenceMaxMs;
            this.SingleSpeakerPerCaption = singleSpeakerPerCaption;
            this.SoundThreshold = soundThreshold;
            this.SoundTokensByCaption = soundTokensByCaption;
            this.SoundTokensByLine = soundTokensByLine;
            this.SoundTokensByCaptionList = soundTokensByCaptionList;
            this.SoundTokensByLineList = soundTokensByLineList;
            this.SpeakerOnNewLine = speakerOnNewLine;
            this.SrtFormat = srtFormat;
            this.SrtPrintCaptionNumbers = srtPrintCaptionNumbers;
            this.StripSquareBrackets = stripSquareBrackets;
            this.Utf8Mark = utf8_mark;
        }
    }

    /* All properties are nullable and are null by default */
    /* If the property is null during the ToQuery() method call, it is left out from the resulting query string */
    public class PerformTranscriptionOptions : Options
    {
        [QueryName("customer_approval_steps")]
        public CustomerApprovalSteps? CustomerApprovalSteps { get; set; }
        [QueryName("customer_approval_tool")]
        public CustomerApprovalTools? CustomerApprovalTool { get; set; }
        [QueryName("custom_metadata")]
        public string CustomMetadata { get; set; }
        [QueryName("notes")]
        public string Notes { get; set; }
        [QueryName("return_iwp")]
        public List<Fidelity> ReturnIwp { get; set; }
        [QueryName("speaker_id")]
        public bool? SpeakerId { get; set; }

        public PerformTranscriptionOptions(CustomerApprovalSteps? customerApprovalSteps=null,
                                           CustomerApprovalTools? customerApprovalTool=null,
                                           string customMetadata = null,
                                           string notes = null,
                                           List<Fidelity> returnIwp = null,
                                           bool? speakerId = null)
        {
            this.CustomerApprovalSteps = customerApprovalSteps;
            this.CustomerApprovalTool = customerApprovalTool;
            this.CustomMetadata = customMetadata;
            this.Notes = notes;
            this.ReturnIwp = returnIwp;
            this.SpeakerId = speakerId;
        }
    }

    // TODO: remove this class, add-O option to command line interface
    public class QueryOptions : Options
    {
        public string QueryString { get; set; }

        public QueryOptions(string queryString = null)
        {
            this.QueryString = queryString;
        }

        public override Dictionary<string, string> GetDictionary()
        {
            if (this.QueryString == null || this.QueryString.Length == 0)
            {
                return new Dictionary<string, string>(); // empty dictionary
            }
            Dictionary<string, string> dict = new Dictionary<string, string>();
            string[] str = this.QueryString.Split("=".ToCharArray(), 2);
            dict.Add(str[0], str[1]);
            return dict;
        }

        public virtual string ToQuery()
        {
            return this.QueryString;
        }
    }

    /* Property Name in this class is used as key in key-value pairs
     * that make up query strings. Value - is the value of a property
     * this attribute is applied to. */
    [AttributeUsage(AttributeTargets.Property)]
    public class QueryName : Attribute
    {
        public string Name;

        public QueryName(string name)
        {
            this.Name = name;
        }
    }
}

File: cielo24_Ruby/cielo24_gem/lib/cielo24/options.rb
module Cielo24
  class BaseOptions

    def get_hash
      hash = {}
      self.instance_variables.each{ |var|
        value = self.instance_variable_get(var)
        if !value.nil?
          hash[var.to_s.delete("@")] = self.instance_variable_get(var)
        end
      }
      return hash
    end

    def to_query
      hash = get_hash
      array = Array.new()
      hash.each do |key, value|
        array.push(key + '=' + value.to_s)
      end
      return array.join("&")
    end

    def populate_from_key_value_pair(key, value)
      # Iterate over instance variables
      self.instance_variables.each{ |var|
        # key gets converted to_s because it can be a Symbol
        if var.to_s.delete("@") == key.to_s
          self.instance_variable_set(var, value)
          break
        end
      }
    end

    def populate_from_hash(hash)
      hash.each do |key, value|
        populate_from_key_value_pair(key, value)
      end
    end
  end

  class CommonOptions < BaseOptions

    attr_accessor :elementlist_version
    attr_accessor :emit_speaker_change_token_as
    attr_accessor :mask_profanity
    attr_accessor :remove_disfluencies
    attr_accessor :remove_sounds_list
    attr_accessor :remove_sound_references
    attr_accessor :replace_slang
    attr_accessor :sound_boundaries

    def initialize(elementlist_version=nil,
                   emit_speaker_change_token_as=nil,
                   mask_profanity=nil,
                   remove_disfluencies=nil,
                   remove_sounds_list=nil,
                   remove_sound_references=nil,
                   replace_slang=nil,
                   sound_boundaries=nil)
      @elementlist_version = elementlist_version
      @emit_speaker_change_token_as = emit_speaker_change_token_as
      @mask_profanity = mask_profanity
      @remove_disfluencies = remove_disfluencies
      @remove_sounds_list = remove_sounds_list
      @remove_sound_references = remove_sound_references
      @replace_slang = replace_slang
      @sound_boundaries = sound_boundaries
    end
  end

  class TranscriptionOptions < CommonOptions

    attr_accessor :create_paragraphs
    attr_accessor :newlines_after_paragraph
    attr_accessor :newlines_after_sentence
    attr_accessor :timecode_every_paragraph
    attr_accessor :timecode_format
    attr_accessor :timecode_interval
    attr_accessor :timecode_offset

    def initialize(create_paragraphs=nil,
                   newlines_after_paragraph=nil,
                   newlines_after_sentence=nil,
                   timecode_every_paragraph=nil,
                   timecode_format=nil,
                   timecode_interval=nil,
                   timecode_offset=nil)
      @create_paragraphs = create_paragraphs
      @newlines_after_paragraph = newlines_after_paragraph
      @newlines_after_sentence = newlines_after_sentence
      @timecode_every_paragraph = timecode_every_paragraph
      @timecode_format = timecode_format
      @timecode_interval = timecode_interval
      @timecode_offset = timecode_offset
    end
  end

  class CaptionOptions < CommonOptions

    attr_accessor :build_url
    attr_accessor :caption_words_min
    attr_accessor :caption_by_sentence
    attr_accessor :characters_per_caption_line
    attr_accessor :dfxp_header
    attr_accessor :disallow_dangling
    attr_accessor :display_effects_speaker_as
    attr_accessor :display_speaker_id
    attr_accessor :force_case
    attr_accessor :include_dfxp_metadata
    attr_accessor :layout_target_caption_length_ms
    attr_accessor :line_break_on_sentence
    attr_accessor :line_ending_format
    attr_accessor :lines_per_caption
    attr_accessor :maximum_caption_duration
    attr_accessor :merge_gap_interval
    attr_accessor :minimum_caption_length_ms
    attr_accessor :minimum_gap_between_captions_ms
    attr_accessor :qt_seamless
    attr_accessor :silence_max_ms
    attr_accessor :single_speaker_per_caption
    attr_accessor :sound_threshold
    attr_accessor :sound_tokens_by_caption
    attr_accessor :sound_tokens_by_line
    attr_accessor :sound_tokens_by_caption_list
    attr_accessor :sound_tokens_by_line_list
    attr_accessor :speaker_on_new_line
    attr_accessor :srt_format
    attr_accessor :strip_square_brackets
    attr_accessor :utf8_mark

    def initialize(build_url=nil,
                   caption_words_min=nil,
                   caption_by_sentence=nil,
                   characters_per_caption_line=nil,
                   dfxp_header=nil,
                   disallow_dangling=nil,
                   display_effects_speaker_as=nil,
                   display_speaker_id=nil,
                   force_case=nil,
                   include_dfxp_metadata=nil,
                   layout_target_caption_length_ms=nil,
                   line_break_on_sentence=nil,
                   line_ending_format=nil,
                   lines_per_caption=nil,
                   maximum_caption_duration=nil,
                   merge_gap_interval=nil,
                   minimum_caption_length_ms=nil,
                   minimum_gap_between_captions_ms=nil,
                   qt_seamless=nil,
                   silence_max_ms=nil,
                   single_speaker_per_caption=nil,
                   sound_threshold=nil,
                   sound_tokens_by_caption=nil,
                   sound_tokens_by_line=nil,
                   sound_tokens_by_caption_list=nil,
                   sound_tokens_by_line_list=nil,
                   speaker_on_new_line=nil,
                   srt_format=nil,
                   strip_square_brackets=nil,
                   utf8_mark=nil)
      @build_url = build_url
      @caption_words_min = caption_words_min
      @caption_by_sentence = caption_by_sentence
      @characters_per_caption_line = characters_per_caption_line
      @dfxp_header = dfxp_header
      @disallow_dangling = disallow_dangling
      @display_effects_speaker_as = display_effects_speaker_as
      @display_speaker_id = display_speaker_id
      @force_case = force_case
      @include_dfxp_metadata = include_dfxp_metadata
      @layout_target_caption_length_ms = layout_target_caption_length_ms
      @line_break_on_sentence = line_break_on_sentence
      @line_ending_format = line_ending_format
      @lines_per_caption = lines_per_caption
      @maximum_caption_duration = maximum_caption_duration
      @merge_gap_interval = merge_gap_interval
      @minimum_caption_length_ms = minimum_caption_length_ms
      @minimum_gap_between_captions_ms = minimum_gap_between_captions_ms
      @qt_seamless = qt_seamless
      @silence_max_ms = silence_max_ms
      @single_speaker_per_caption = single_speaker_per_caption
      @sound_threshold = sound_threshold
      @sound_tokens_by_caption = sound_tokens_by_caption
      @sound_tokens_by_line = sound_tokens_by_line
      @sound_tokens_by_caption_list = sound_tokens_by_caption_list
      @sound_tokens_by_line_list = sound_tokens_by_line_list
      @speaker_on_new_line = speaker_on_new_line
      @srt_format = srt_format
      @strip_square_brackets = strip_square_brackets
      @utf8_mark = utf8_mark
    end
  end

  class PerformTranscriptionOptions < BaseOptions

    attr_accessor :customer_approval_steps
    attr_accessor :customer_approval_tool
    attr_accessor :custom_metadata
    attr_accessor :generate_media_intelligence_for_iwp
    attr_accessor :notes
    attr_accessor :return_iwp
    attr_accessor :speaker_id

    def initialize(customer_approval_steps=nil,
                   customer_approval_tool=nil,
                   custom_metadata=nil,
                   generate_media_intelligence_for_iwp=nil,
                   notes=nil,
                   return_iwp=nil,
                   speaker_id=nil)
      @customer_approval_steps = customer_approval_steps
      @customer_approval_tool = customer_approval_tool
      @custom_metadata = custom_metadata
      @generate_media_intelligence_for_iwp = generate_media_intelligence_for_iwp
      @notes = notes
      @return_iwp = return_iwp
      @speaker_id = speaker_id
    end
  end

  class JobListOptions < BaseOptions

    attr_accessor :CreationDateFrom
    attr_accessor :CreationDateTo
    attr_accessor :StartDateFrom
    attr_accessor :StartDateTo
    attr_accessor :DueDateFrom
    attr_accessor :DueDateTo
    attr_accessor :CompleteDateFrom
    attr_accessor :CompleteDateTo
    attr_accessor :JobStatus
    attr_accessor :Fidelity
    attr_accessor :Priority
    attr_accessor :TurnaroundTimeHoursFrom
    attr_accessor :TurnaroundTimeHoursTo
    attr_accessor :username

    def initialize(creation_date_from=nil,
                   creation_date_to=nil,
                   start_date_from=nil,
                   start_date_to=nil,
                   due_date_from=nil,
                   due_date_to=nil,
                   complete_date_from=nil,
                   complete_date_to=nil,
                   job_status=nil,
                   fidelity=nil,
                   priority=nil,
                   turnaround_time_hours_from=nil,
                   turnaround_time_hours_to=nil,
                   sub_account=nil)
      @CreationDateFrom = creation_date_from
      @CreationDateTo = creation_date_to
      @StartDateFrom = start_date_from
      @StartDateTo = start_date_to
      @DueDateFrom = due_date_from
      @DueDateTo = due_date_to
      @CompleteDateFrom = complete_date_from
      @CompleteDateTo = complete_date_to
      @JobStatus = job_status
      @Fidelity = fidelity
      @Priority = priority
      @TurnaroundTimeHoursFrom = turnaround_time_hours_from
      @TurnaroundTimeHoursTo = turnaround_time_hours_to
      @username = sub_account
    end
  end
end